File: 프로그래머스/1/340199. ［PCCE 기출문제］ 9번 ／ 지폐 접기/Tae Seong Hong/［PCCE 기출문제］ 9번 ／ 지폐 접기.cpp
#include <string>
#include <vector>

using namespace std;

int max(int a, int b) {
    if(a>b)
        return a;
    else
        return b;
}

int min(int a, int b) {
    if(a>b)
        return b;
    else
        return a;
}

int solution(vector<int> wallet, vector<int> bill) {
    int answer = 0;
    
    while(1) {
        int min_bill = min(bill[0], bill[1]);
        int max_bill = max(bill[0], bill[1]);
        int min_wallet = min(wallet[0], wallet[1]);
        int max_wallet = max(wallet[0], wallet[1]);
        if(min_bill > min_wallet || max_bill > max_wallet) {
            if(bill[0] > bill[1])
                bill[0] /= 2;
            else {
                bill[1] /= 2;
            }
            answer++;
            continue;
        }
        break;
    }
    
    return answer;
}

/*
길이가 긴 쪽을 반으로
홀수(odd) 였다면 소수점 이하는 버림
90도 돌려서 사용가능하면 끝
*/

File: 프로그래머스/1/340199. ［PCCE 기출문제］ 9번 ／ 지폐 접기/Hyun-suck Lee/［PCCE 기출문제］ 9번 ／ 지폐 접기.java
class Solution {
    public int solution(int[] wallet, int[] bill) {
        int answer = 0;
        this.order(wallet);
        
        while (true) {
            order(bill);
            if (checkFit(wallet, bill)) break;
            
            bill[1] >>= 1;
            answer++;
        }
        
        return answer;
    }

    private boolean checkFit(int[] space, int[] thing) {
        return space[0] >= thing[0] && space[1] >= thing[1];
    }
    
    // 원소 개수 2개로 정해졌으므로 편하게 스왑으로
    private void order(int[] target) {
        if (target[0] > target[1]) {
            int temp = target[0];
            target[0] = target[1];
            target[1] = temp;
        }
    }
}